storage2/src/storage.rs

```diff
@@ -0,0 +1,227 @@
+use std::{path::PathBuf, sync::Arc};
+
+use ::metrics::gauge;
+use anyhow::Result;
+use futures::future::BoxFuture;
+use jmt::{
+    storage::{Node, NodeBatch, NodeKey, TreeReader, TreeWriter},
+    WriteOverlay,
+};
+use rocksdb::{Options, DB};
+use tokio::sync::RwLock;
+use tracing::Span;
+
+use penumbra_tct as tct;
+
+use crate::{metrics, State};
+
+#[derive(Clone, Debug)]
+pub struct Storage(Arc<DB>);
+
+impl Storage {
+    pub async fn load(path: PathBuf) -> Result<Self> {
+        let span = Span::current();
+        tokio::task::Builder::new()
+            .name("open_rocksdb")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    tracing::info!(?path, "opening rocksdb");
+                    let mut opts = Options::default();
+                    opts.create_if_missing(true);
+                    opts.create_missing_column_families(true);
+
+                    Ok(Self(Arc::new(DB::open_cf(&opts, path, ["jmt", "nct"])?)))
+                })
+            })
+            .unwrap()
+            .await
+            .unwrap()
+    }
+
+    /// Returns the latest version (block height) of the tree recorded by the
+    /// `Storage`, or `None` if the tree is empty.
+    pub async fn latest_version(&self) -> Result<Option<jmt::Version>> {
+        Ok(self
+            .get_rightmost_leaf()
+            .await?
+            .map(|(node_key, _)| node_key.version()))
+    }
+
+    /// Returns a new [`State`] on top of the latest version of the tree.
+    pub async fn state(&self) -> Result<State> {
+        // If the tree is empty, use PRE_GENESIS_VERSION as the version,
+        // so that the first commit will be at version 0.
+        let version = self
+            .latest_version()
+            .await?
+            .unwrap_or(WriteOverlay::<Storage>::PRE_GENESIS_VERSION);
+
+        tracing::debug!("creating state for version {}", version);
+        Ok(Arc::new(RwLock::new(WriteOverlay::new(
+            self.clone(),
+            version,
+        ))))
```

Comment:
This implementation isn't correct, because a `State` isn't an Arc<... any more, so we should delete it until we define what a `State` actually is and what data it consists of.